libpod/sqlite_state.go

```diff
@@ -0,0 +1,2535 @@
+// container that the database is aware of (IE, were added via AddExecSession).
+func (s *SQLiteState) GetContainerExecSessions(ctr *Container) ([]string, error) {
+	if !s.valid {
+		return nil, define.ErrDBClosed
+	}
+
+	if !ctr.valid {
+		return nil, define.ErrCtrRemoved
+	}
+
+	rows, err := s.conn.Query("SELECT ID FROM ContainerExecSession WHERE ContainerID=?;", ctr.ID())
+	if err != nil {
+		return nil, fmt.Errorf("querying container %s exec sessions: %w", ctr.ID(), err)
+	}
+	defer rows.Close()
+
+	var sessions []string
+	for rows.Next() {
+		var session string
+		if err := rows.Scan(&session); err != nil {
+			return nil, fmt.Errorf("scanning container %s exec sessions row: %w", ctr.ID(), err)
+		}
+		sessions = append(sessions, session)
+	}
+
+	return sessions, nil
+}
+
+// RemoveContainerExecSessions removes all exec sessions attached to a given
+// container.
+func (s *SQLiteState) RemoveContainerExecSessions(ctr *Container) (defErr error) {
+	if !s.valid {
+		return define.ErrDBClosed
+	}
+
+	if !ctr.valid {
+		return define.ErrCtrRemoved
+	}
+
+	tx, err := s.conn.Begin()
+	if err != nil {
+		return fmt.Errorf("beginning container %s exec session removal transaction: %w", ctr.ID(), err)
+	}
+	defer func() {
+		if defErr != nil {
+			if err := tx.Rollback(); err != nil {
+				logrus.Errorf("Rolling back transaction to remove container %s exec sessions: %v", ctr.ID(), err)
+			}
+		}
+	}()
+
+	if _, err := tx.Exec("DELETE FROM ContainerExecSessions WHERE ContainerID=?;", ctr.ID()); err != nil {
+		return fmt.Errorf("removing container %s exec sessions from database: %w", ctr.ID(), err)
+	}
+
+	if err := tx.Commit(); err != nil {
+		return fmt.Errorf("committing container %s exec session removal: %w", ctr.ID(), err)
+	}
+
+	return nil
+}
+
+// RewriteContainerConfig rewrites a container's configuration.
+// DO NOT USE TO: Change container dependencies, change pod membership, change
+// container ID.
+// WARNING: This function is DANGEROUS. Do not use without reading the full
+// comment on this function in state.go.
+// TODO: Once BoltDB is removed, this can be combined with SafeRewriteContainerConfig.
+func (s *SQLiteState) RewriteContainerConfig(ctr *Container, newCfg *ContainerConfig) error {
+	if !s.valid {
+		return define.ErrDBClosed
+	}
+
+	if !ctr.valid {
+		return define.ErrCtrRemoved
+	}
+
+	return s.rewriteContainerConfig(ctr, newCfg)
+}
+
+// SafeRewriteContainerConfig rewrites a container's configuration in a more
+// limited fashion than RewriteContainerConfig. It is marked as safe to use
+// under most circumstances, unlike RewriteContainerConfig.
+// DO NOT USE TO: Change container dependencies, change pod membership, change
+// locks, change container ID.
+// TODO: Once BoltDB is removed, this can be combined with RewriteContainerConfig.
+func (s *SQLiteState) SafeRewriteContainerConfig(ctr *Container, oldName, newName string, newCfg *ContainerConfig) error {
```

Comment:
same question wrt "dangerous" methods

Reply:
It's not accessible outside of Libpod because the DB is never passed out of libpod.